support/ebpf/beam_tracer.ebpf.c

```diff
@@ -2,19 +2,131 @@
 #include "tracemgmt.h"
 #include "types.h"
 
+// The number of frames to unwind per frame-unwinding eBPF program.
+#define BEAM_FRAMES_PER_PROGRAM 8
+
+// The max number of loops to unroll when searching for the correct CodeHeader.
+// Should be log base 2 of a reasonable number of modules to binary-search through.
+#define BEAM_CODE_HEADER_SEARCH_ITERATIONS 16
+
+// The max number of loops to unroll when scanning the stack from for continuation pointers
+#define BEAM_STACK_FRAME_SCAN_ITERATIONS 16
+
 struct beam_procs_t {
   __uint(type, BPF_MAP_TYPE_HASH);
   __type(key, pid_t);
   __type(value, BEAMProcInfo);
   __uint(max_entries, 256);
 } beam_procs SEC(".maps");
 
+// We assume this Range struct is stable so we can read it all at once directly.
+// If it were to change in the future, we'd need to change the way this struct is read.
+// https://github.com/erlang/otp/blob/OTP-27.2.4/erts/emulator/beam/beam_ranges.c#L31-L34
+typedef struct BEAMRangeEntry {
+  u64 start;
+  u64 end;
+} BEAMRangeEntry;
+
+// We assume this ranges struct will at least have these two values as the first two fields.
+// That allows us to directly read them efficiently at once, but it could break in in the future.
+// https://github.com/erlang/otp/blob/OTP-27.2.4/erts/emulator/beam/beam_ranges.c#L56-L61
+typedef struct BEAMRangesInfo {
+  u64 modules;
+  u64 n;
+} BEAMRangesInfo;
+
+typedef struct BEAMRangesSearchCache {
+  BEAMRangesInfo info;
+  BEAMRangeEntry first, mid, last;
+} BEAMRangesSearchCache;
+
+static EBPF_INLINE ErrorCode
+unwind_one_beam_frame(PerCPURecord *record, BEAMProcInfo *info, BEAMRangesSearchCache *ranges)
+{
+  UnwindState *state = &record->state;
+  Trace *trace       = &record->trace;
+  u64 pc             = state->pc;
+
+  if (pc < ranges->first.start || pc > ranges->last.end) {
+    return ERR_BEAM_PC_INVALID;
+  }
+
+  unwinder_mark_nonleaf_frame(state);
```

Comment:
This should be moved to where the state is updated with unwound frame.

Resetting this here will give wrong flag for the first frame.

The frame pointer unwind helper also does this already at https://github.com/open-telemetry/opentelemetry-ebpf-profiler/blob/2aba5ae70a99e1cfc0cb0a013a588b5ec731cf1f/support/ebpf/tracemgmt.h#L318 so only the other path needs it. Or perhaps it gets optimized out if this put in the common path.